### PR TITLE
Basic Nitro Integration (Version 0)

### PR DESCRIPTION
**Pull request by lalexgap**
Fixes https://github.com/statechannels/boost/issues/5
Fixes https://github.com/statechannels/boost/issues/4

This implements a basic integration of nitro payments into the boost http endpoint.

If nitro payments are enabled (controlled by a new flag added to boost http, `--nitro-enabled`) boost http now checks for a payment before serving up a payload or piece. When requesting a file with nitro enabled the channelId to use must be passsed via a query parameters:
 ```http://localhost:7777/ipfs/bafk2bzaceapnitekx4sp3mtitqatm5zpxn6nvjicwtltomttrlof65wlcfjpa?channelId=0x0596b064913deae8892f6c848efaaf144746e91a8d8bc0abd67d05f0a1285c03```

Currently there's no real "accounting" so the payment check simply checks if the `payer` has paid `10` on the given channel.

### CODE REVIEW

File: cmd/booster-http/server.go [hunk 1]
```diff
@@ -164,6 +187,27 @@ func (s *HttpServer) handleByPieceCid(w http.ResponseWriter, r *http.Request) {
...
+
+		if (chId == types.Destination{}) {
+			writeError(w, r, http.StatusPaymentRequired, "a valid channel id must be provided")
+			return
+		}
+		// TODO: Allow this to be configurable
+		expectedPaymentAmount := big.NewInt(10)
+
+		if hasPaid := checkPaymentChannelBalance(s.nitroRpcClient, chId, expectedPaymentAmount); !hasPaid {
+			writeError(w, r, http.StatusPaymentRequired, "payment required")
```
Comment (geoknee) [inline cmd/booster-http/server.go:207]: Probably out of scope for this PR, but I think with this model, one payment would unlock an unlimited number of retrievals. Longer term we need a better model.

File: cmd/booster-http/server.go [hunk 2]
```diff
@@ -66,11 +73,26 @@ type HttpServerOptions struct {
...
 		opts = &HttpServerOptions{ServePieces: true}
 	}
-	return &HttpServer{path: path, listenAddr: listenAddr, port: port, api: api, opts: *opts, idxPage: parseTemplate(*opts)}
+	var rpcClient *nrpc.RpcClient
+	var err error
+	if nitroOpts != nil && nitroOpts.Enabled {
+
+		rpcClient, err = nrpc.NewHttpRpcClient(nitroOpts.Endpoint)
+		if err != nil {
+			panic(err)
```
Comment (geoknee) [inline cmd/booster-http/server.go:91]: Probably ok to panic here given we are inside a constructor.

File: cmd/booster-http/server.go [hunk 3]
```diff
@@ -155,6 +177,7 @@ func (s *HttpServer) handleByPieceCid(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
+	// TODO: Update to parse out multiple url params
```
Comment (geoknee) [inline cmd/booster-http/server.go:180]: Let's track in a new issue?

File: cmd/booster-http/util.go [hunk 4]
```diff
@@ -9,3 +16,29 @@ func addCommas(count uint64) string {
 	}
 	return str
 }
+
+// checkPaymentChannelBalance checks a payment channel balance and returns true if the AmountPaid is greater than the expected amount
+func checkPaymentChannelBalance(rpcClient *rpc.RpcClient, paymentChannelId types.Destination, expectedAmount *big.Int) bool {
+	if rpcClient == nil {
+		panic("the rpcClient is nil")
```
Comment (geoknee) [inline cmd/booster-http/util.go:23]: I guess this panic is justified by "it should never happen"? Perhaps we should warn about the panic in the godoc?

File: cmd/booster-http/gateway_handler.go [hunk 5]
```diff
@@ -22,9 +29,11 @@ func newGatewayHandler(gw *BlocksGateway, supportedFormats []string) http.Handle
 		fmtsMap[f] = struct{}{}
 	}
 
+	// TODO: For the integration demo, we need to allow CORS requests to the gateway.
 	return &gatewayHandler{
-		gwh:              gateway.NewHandler(gateway.Config{Headers: headers}, gw),
+		gwh:              &corsHandler{gateway.NewHandler(gateway.Config{Headers: headers}, gw)},
```
Comment (geoknee) [inline cmd/booster-http/gateway_handler.go:34]: Definitely out of scope: I would prefer a variable name like `h` over `gwh`. Seems confusing to have multiplte "gatewayhandler" types nested like this.

File: go.mod [hunk 6]
```diff
@@ -347,25 +351,59 @@ require (
 
 require (
 	github.com/NYTimes/gziphandler v1.1.1
+	github.com/ethereum/go-ethereum v1.11.4
```
Comment (geoknee) [inline go.mod:354]: As mentioned elsewhere, it would be nice to not need to depend on this.

File: cmd/booster-http/http_test.go [hunk 7]
```diff
@@ -109,7 +109,7 @@ func TestHttpInfo(t *testing.T) {
 
 	// Create a new mock Http server
 	ctrl := gomock.NewController(t)
-	httpServer := NewHttpServer("", "0.0.0.0", 7777, mocks_booster_http.NewMockHttpServerApi(ctrl), nil)
```
Comment (geoknee) [inline cmd/booster-http/http_test.go:112]: Can we add a test which exercises the new code somehow?

**Review by geoknee — APPROVED** [review]
This is a great first step @lalexgap !

File: cmd/booster-http/server.go (hunk 3 above)
Comment (lalexgap) [reply cmd/booster-http/server.go:180]: Oops, I think this was an old TODO. The existing logic **can** parse out multiple params now.

File: cmd/booster-http/server.go (hunk 1 above)
Comment (lalexgap) [reply cmd/booster-http/server.go:207]: Yup! This doesn't check if you paid for a file; it checks that you've paid at least 10 on some channel (once).

File: cmd/booster-http/server.go [hunk 8]
```diff
@@ -66,11 +73,26 @@ type HttpServerOptions struct {
...
+}
+
+func NewHttpServer(path string, listenAddr string, port int, api HttpServerApi, opts *HttpServerOptions, nitroOpts *NitroOptions) *HttpServer {
 	if opts == nil {
 		opts = &HttpServerOptions{ServePieces: true}
 	}
-	return &HttpServer{path: path, listenAddr: listenAddr, port: port, api: api, opts: *opts, idxPage: parseTemplate(*opts)}
+	var rpcClient *nrpc.RpcClient
+	var err error
+	if nitroOpts != nil && nitroOpts.Enabled {
```
[the remark this answers is not in the document]
Comment (lalexgap) [reply cmd/booster-http/server.go:87]: Yup that could work. Although I suspect that we'll want to provide more options/configuration in the future so we'll probably want something like `NitroOptions` eventually.

File: go.mod (hunk 6 above)
Comment (geoknee) [reply go.mod:354]: yeah I think an indirect reference is OK / unavoidable.

File: cmd/booster-http/server.go [hunk 9]
```diff
@@ -164,6 +187,27 @@ func (s *HttpServer) handleByPieceCid(w http.ResponseWriter, r *http.Request) {
...
+			writeError(w, r, http.StatusPaymentRequired, "a valid channel id must be provided")
+			return
+		}
+		rawChId := params.Get("channelId")
+		chId := types.Destination(common.HexToHash(rawChId))
+
+		if (chId == types.Destination{}) {
+			writeError(w, r, http.StatusPaymentRequired, "a valid channel id must be provided")
+			return
+		}
```
[the remark this answers is not in the document]
Comment (geoknee) [reply cmd/booster-http/server.go:202]: Yep all good for demo purposes. Over the medium term, let's try and expedite the RPC error handling on the go-nitro side of things.

File: cmd/booster-http/server.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (geoknee) [reply cmd/booster-http/server.go:87]: https://github.com/statechannels/boost/issues/8

File: cmd/booster-http/server.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (geoknee) [reply cmd/booster-http/server.go:202]: https://github.com/statechannels/boost/issues/8

[the remark this answers is not in the document]
Comment (bitwiseguy) [reply cmd/booster-http/server.go:202]: > Could we leave this validation up to the go-nitro node? As in, we'll send the query with whatever channelId param, and catch the error coming back.
> 
> That way, we make this integration leaner and don't need to import `go-nitro` types.
> 
> If we design our rpc client API well, I think consumers should not need to import anything else (especially `go-ethereum`).

@geoknee in this comment are you suggesting that in the [go-nitro rpc client](https://github.com/statechannels/go-nitro/blob/main/rpc/client.go) we should only use standard types (opposed to custom go-nitro types) for both inputs and outputs for those methods? I just want to clarify before diving too deep into implementation

[the remark this answers is not in the document]
Comment (geoknee) [reply cmd/booster-http/server.go:202]: Let's discuss today at standup.

Potentially, we could just do this for the inputs to the API (and not the outputs).

Or, we may decide that it's not worth the effort to make such changes. Or indeed that it closes off some nice/important functionality for the consumer.

**Pull request merged**